segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1417 @@
+     */
+    @Getter
+    private int containerId;
+
+    /**
+     * {@link SystemJournal} that logs all changes to system segment layout so that they can be are used during system bootstrap.
+     */
+    @Getter
+    private SystemJournal systemJournal;
+
+    /**
+     * Index of chunks for a segment by their start offsets.
+     */
+    private final ConcurrentHashMap<String, ConcurrentSkipListMap<Long, String>> cachedReadIndex = new ConcurrentHashMap<String, ConcurrentSkipListMap<Long, String>>();
+
+    /**
+     * Creates a new instance of the ChunkStorageManager class.
+     *
+     * @param chunkStorage ChunkStorageProvider instance.
+     * @param executor    An Executor for async operations.
+     * @param config Configuration options for this ChunkStorageManager instance.
+     */
+    public ChunkStorageManager(ChunkStorageProvider chunkStorage, Executor executor, ChunkStorageManagerConfig config) {
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Creates a new instance of the ChunkStorageManager class.
+     *
+     * @param chunkStorage ChunkStorageProvider instance.
+     * @param metadataStore Metadata store.
+     * @param executor    An Executor for async operations.
+     * @param config Configuration options for this ChunkStorageManager instance.
+     */
+    public ChunkStorageManager(ChunkStorageProvider chunkStorage, ChunkMetadataStore metadataStore, Executor executor, ChunkStorageManagerConfig config) {
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Initializes the ChunkStorageManager.
+     *
+     * @param metadataStore Metadata store.
+     * @param containerId container id.
+     * @param systemJournal SystemJournal that keeps track of changes to system segments and helps with bootstrap.
+     * @throws Exception In case of any errors.
+     */
+    public void initialize(int containerId, ChunkMetadataStore metadataStore, SystemJournal systemJournal) throws Exception {
+        this.containerId = containerId;
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.systemJournal = Preconditions.checkNotNull(systemJournal, "systemJournal");
+    }
+
+    /**
+     * Initializes this instance with the given ContainerEpoch.
+     *
+     * @param containerEpoch The Container Epoch to initialize with.
+     */
+    @Override
+    public void initialize(long containerEpoch) {
```

Comment:
The reason for two initialize is that segment container needs storage initialized before it can use table segments and and storage needs table segments to function. To break the circular dependency, the storage is initialized in multiple steps. may be we can have distinct names than multiple overloads. Most of the non final fields are also coming from this multi step initialization.